File: scripts/lib/wiktionary.ts
// Reads the Wiktionary data, and — the actual work — decides which entries to
// keep. That decision lives in code rather than a document because it sets both
// the storage bill and how much noise search has to sift. Every rule is
// separate and counted separately, so the build can report how many entries
// each one removed. A filter nobody can attribute is a filter nobody can argue
// with.
//
// The source is one entry per line, each holding a word, its part of speech and
// its meanings — which maps straight onto a row in our index.
//
// Wiktionary is share-alike licensed. The attribution below is recorded and
// must travel with any index built from these rows.
//
// Two things deliberately NOT filtered on. Rarity: the frequency table we have
// contains none of `petrichor`, `hangry` or `doomscrolling`, so filtering by it
// would remove exactly the words worth adding. Informality: slang and internet
// words are the payload, and dropping them would look like quality control
// while deleting the point of the exercise.

/** The credit that has to travel with every row built from this source. */
export const LICENCE =
  "English Wiktionary, via the Kaikki.org wiktextract extraction. " +
  "Text is CC BY-SA 4.0 / GFDL: attribution and share-alike both apply.";

export const SOURCE_URL =
  "https://kaikki.org/dictionary/English/kaikki.org-dictionary-English.jsonl";

/** Bump this whenever a rule below changes; it gets recorded with each run. */
export const FILTER_VERSION = "rd17.2";

/** The four kinds of word the index holds: nouns, verbs, adjectives, adverbs. */
// The source has many more — names, prepositions, numbers — but nobody
// describes those to a reverse dictionary.
export const KEPT_POS = new Set(["noun", "verb", "adj", "adv"]);

/** Marks an entry that just points at another word rather than meaning something. */
// "Plural of cat" is a dead end with no answer behind it. Slang and internet
// words are deliberately not in this list — they are what we came for.
export const FORM_TAGS = new Set([
  "form-of",
  "alt-of",
  "alternative",
  "abbreviation",
  "initialism",
  "acronym",
  "misspelling",
  "plural",
  "singular",
  "past",
  "participle",
  "comparative",
  "superlative",
  "romanization",
  "pronunciation-spelling",
  "eye-dialect",
  "clipping",
  "ellipsis",
  "contraction",
]);

/** Marks a meaning nobody would be describing today. */
// Long-dead senses are the worst kind of near-miss to have in the index.
// "Rare" is deliberately absent: `petrichor` and `limerence` are rare.
export const DEAD_TAGS = new Set(["obsolete", "archaic", "dated", "dialectal"]);

/** Definitions that are really just cross-references written out in words. */
// The tags above catch most of these, but the source doesn't always set them.
export const SHELL_GLOSS =
  /^(alternative|obsolete|archaic|dated|nonstandard|informal|common|eye)?\s*(spelling|form|misspelling|pronunciation|romanization)\s+of\b/i;

export const INFLECTION_GLOSS =
  /^(plural|singular|past|present|simple past|gerund|inflection|comparative|superlative|third-person|participle|abbreviation|initialism|acronym|clipping|ellipsis|synonym|antonym|obsolete form|used other than|only used in|see\b)/i;

/** The same shape test the main index uses. */
// Multi-word entries are kept: "deja vu" is a legitimate answer.
export function isJunkSurface(word: string): boolean {
  return /^[A-Z]/.test(word) || /[0-9]/.test(word) || /[^A-Za-z '-]/.test(word);
}

/** Shortest definition worth keeping, in characters. */
// "A small fish." says almost nothing against a twelve-word description, and
// there are a great many like it. This is the one judgement call in the filter,
// so it is named and reported rather than buried inside a condition.
export const MIN_GLOSS_CHARS = 20;

/** How many meanings to keep per word and part of speech. */
export const MAX_SENSES_PER_ENTRY = 4;

// ------------------------------------------------------------------ shapes

export type KaikkiSense = {
  glosses?: string[];
  raw_glosses?: string[];
  tags?: string[];
  form_of?: unknown;
  alt_of?: unknown;
};

export type KaikkiEntry = {
  word?: string;
  pos?: string;
  lang_code?: string;
  senses?: KaikkiSense[];
};

/** One row for the index, the same shape the main table uses. */
export type SupplementRow = {
    /** A key that can never clash with the main dictionary's keys. */
  key: string;
  pos: string;
  gloss: string;
    /** Always one word: this source has no groups of synonyms. */
  lemmas: string[];
  source: "wiktionary";
};

/** Why an entry was rejected — one counter per rule, so the filter is auditable. */
export type RejectReason =
  | "lang"
  | "pos"
  | "surface"
  | "form_ref"
  | "form_tag"
  | "dead_tag"
  | "no_gloss"
  | "shell_gloss"
  | "short_gloss"
  | "sense_cap"
  | "duplicate"
  | "already_answerable";

export type FilterCounts = Record<RejectReason | "kept", number>;

/** State the filter carries between entries. */
// The two counters answer different questions — "have I indexed this exact
// meaning already?" and "how many meanings of this word have I emitted?" —
// and merging them once produced thousands of clashing keys.
export type FilterState = {
  seen: Set<string>;
  senseIndex: Map<string, number>;
};

export function emptyState(): FilterState {
  return { seen: new Set(), senseIndex: new Map() };
}

export function emptyCounts(): FilterCounts {
  return {
    lang: 0,
    pos: 0,
    surface: 0,
    form_ref: 0,
    form_tag: 0,
    dead_tag: 0,
    no_gloss: 0,
    shell_gloss: 0,
    short_gloss: 0,
    sense_cap: 0,
    duplicate: 0,
    already_answerable: 0,
    kept: 0,
  };
}

/** Clean up one definition. */
// Uses the plain text, not the version prefixed with "(transitive)" and the
// like: that prefix is a label, not meaning, and indexing it would push the
// word "transitive" into every transitive verb.
export function cleanGloss(sense: KaikkiSense): string {
  const text = (sense.glosses ?? [])[0] ?? "";
  return text.replace(/\s+/g, " ").trim();
}

/** Run the filter over one entry, keeping what survives and tallying the rest. */
// Passing a test for "the index can already answer this" switches between the
// two variants being compared.
export function filterEntry(
  entry: KaikkiEntry,
  counts: FilterCounts,
  state: FilterState,
  isAnswerable?: (word: string) => boolean
): SupplementRow[] {
  const { seen, senseIndex } = state;
  const word = (entry.word ?? "").trim();
  const pos = entry.pos ?? "";

  if (entry.lang_code !== "en") {
    counts.lang++;
    return [];
  }
  if (!KEPT_POS.has(pos)) {
    counts.pos++;
    return [];
  }
  if (!word || isJunkSurface(word)) {
    counts.surface++;
    return [];
  }
  if (isAnswerable?.(word.toLowerCase())) {
    counts.already_answerable++;
    return [];
  }

  const rows: SupplementRow[] = [];
  for (const sense of entry.senses ?? []) {
    if (sense.form_of || sense.alt_of) {
      counts.form_ref++;
      continue;
    }
    const tags = sense.tags ?? [];
    if (tags.some((t) => FORM_TAGS.has(t))) {
      counts.form_tag++;
      continue;
    }
    if (tags.some((t) => DEAD_TAGS.has(t))) {
      counts.dead_tag++;
      continue;
    }
    const gloss = cleanGloss(sense);
    if (!gloss) {
      counts.no_gloss++;
      continue;
    }
    if (SHELL_GLOSS.test(gloss) || INFLECTION_GLOSS.test(gloss)) {
      counts.shell_gloss++;
      continue;
    }
    if (gloss.length < MIN_GLOSS_CHARS) {
      counts.short_gloss++;
      continue;
    }
    if (rows.length >= MAX_SENSES_PER_ENTRY) {
      counts.sense_cap++;
      continue;
    }
    // The same word appears under several origins, each its own entry, and those
    // routinely repeat a definition word for word.
    const fingerprint = `${word.toLowerCase()} ${pos} ${gloss.toLowerCase()}`;
    if (seen.has(fingerprint)) {
      counts.duplicate++;
      continue;
    }
    seen.add(fingerprint);

    // Numbered across all of a word's entries, not restarted within each one.
    // Otherwise `cat` the animal and `cat` the command both claim the same key —
    // untidy in a test file, but silent data loss on the way into the database,
    // where the second row would quietly overwrite the first.
    const slug = `${word.toLowerCase().replace(/ /g, "_")}:${pos}`;
    const n = senseIndex.get(slug) ?? 0;
    senseIndex.set(slug, n + 1);

    rows.push({
      key: `wikt:${slug}:${n}`,
      pos,
      gloss,
      lemmas: [word],
      source: "wiktionary",
    });
    counts.kept++;
  }
  return rows;
}

File: training/rdlib/wiktionary.py
"""
Reading the Wiktionary download, FOR TRAINING EXAMPLES ONLY.

Read this before using it. This is deliberately NOT a copy of the TypeScript
filter, and must never be used to reproduce the shipped index. That filter is
what built the entries currently in the live table, its per-rule counts are a
committed record, and a second version of it would be a real drift hazard for a
real production table.

What this does is much narrower: pull (word, part of speech, definition) out of
the raw file so training examples can be built. Nothing here decides what gets
indexed, so nothing here can silently change what users search.

It does mirror the same categorical exclusions -- pointers to other words, dead
usages, junk spellings, a minimum length -- because those are as right for
training data as for index rows: an example whose "definition" is "plural of
cat" teaches nothing. Where it differs, it is stricter, never looser.

The source is a multi-gigabyte download, fetched separately, and share-alike
licensed.
"""

from __future__ import annotations

import json
import os
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Iterator

# --- mirrored from scripts/lib/wiktionary.ts --------------------------------

KEPT_POS = frozenset({"noun", "verb", "adj", "adv"})

# Tags marking a sense as a pointer to another word rather than a meaning.
FORM_TAGS = frozenset(
    {
        "form-of", "alt-of", "alternative", "abbreviation", "initialism",
        "acronym", "misspelling", "plural", "singular", "past", "participle",
        "comparative", "superlative", "romanization", "pronunciation-spelling",
        "eye-dialect", "clipping", "ellipsis", "contraction",
    }
)

# Usages nobody would be describing. "Rare" is deliberately absent: `petrichor`
# and `limerence` are rare in exactly that sense.
DEAD_TAGS = frozenset({"obsolete", "archaic", "dated", "dialectal"})

SHELL_GLOSS = re.compile(
    r"^(alternative|obsolete|archaic|dated|nonstandard|informal|common|eye)?\s*"
    r"(spelling|form|misspelling|pronunciation|romanization)\s+of\b",
    re.IGNORECASE,
)

INFLECTION_GLOSS = re.compile(
    r"^(plural|singular|past|present|simple past|gerund|inflection|comparative|"
    r"superlative|third-person|participle|abbreviation|initialism|acronym|"
    r"clipping|ellipsis|synonym|antonym|obsolete form|used other than|"
    r"only used in|see\b)",
    re.IGNORECASE,
)

# Multi-word entries are kept on purpose -- `deja vu` and `stiff upper lip` are
# legitimate answers.
_JUNK_SURFACE = re.compile(r"^[A-Z]|[0-9]|[^A-Za-z '-]")

MIN_GLOSS_CHARS = 20
MAX_SENSES_PER_ENTRY = 4

# Wiktextract normalises POS to these; map to WordNet's four.
_POS_MAP = {"noun": "noun", "verb": "verb", "adj": "adj", "adv": "adv"}

_WS = re.compile(r"\s+")


def source_dir() -> Path:
    return Path(os.environ.get("RD_SOURCE_DIR", Path.home() / "rd_sources"))


def kaikki_path() -> Path:
    return source_dir() / "kaikki-english.jsonl"


def is_junk_surface(word: str) -> bool:
    return bool(_JUNK_SURFACE.search(word))


def clean_gloss(sense: dict) -> str:
    """
    The first definition, tidied up.

    Uses the plain text, not the version prefixed with "(transitive)" and the
    like: that prefix is a label, not meaning.
    """
    glosses = sense.get("glosses") or []
    return _WS.sub(" ", glosses[0]).strip() if glosses else ""


@dataclass(frozen=True)
class WiktSense:
    word: str
    pos: str
    gloss: str


def iter_senses(
    path: Path | None = None,
    *,
    limit_entries: int | None = None,
    progress_every: int | None = 500_000,
) -> Iterator[WiktSense]:
    """
    Read surviving entries one at a time; the file is far too big to load whole.

    A full pass takes a few minutes. Pass `limit_entries` while developing.
    """
    path = path or kaikki_path()
    if not path.exists():
        raise FileNotFoundError(
            f"{path} not found. Fetch it with `npm run supplement:fetch` from "
            "the repo root (~3.2 GB), or set RD_SOURCE_DIR."
        )

    with path.open(encoding="utf-8") as fh:
        for n, line in enumerate(fh):
            if limit_entries is not None and n >= limit_entries:
                return
            if progress_every and n and n % progress_every == 0:
                print(f"  read {n:,} entries", end="\r")

            line = line.strip()
            if not line:
                continue
            try:
                entry = json.loads(line)
            except json.JSONDecodeError:
                                # The file has the odd broken line; skipping is correct and the
                                # count is immaterial at this scale.
                continue

            if entry.get("lang_code") != "en":
                continue
            pos = _POS_MAP.get(entry.get("pos", ""))
            if pos is None:
                continue
            word = entry.get("word") or ""
            if not word or is_junk_surface(word):
                continue

            kept = 0
            for sense in entry.get("senses") or []:
                if kept >= MAX_SENSES_PER_ENTRY:
                    break
                tags = set(sense.get("tags") or [])
                if tags & FORM_TAGS or tags & DEAD_TAGS:
                    continue
                if sense.get("form_of") or sense.get("alt_of"):
                    continue
                gloss = clean_gloss(sense)
                if len(gloss) < MIN_GLOSS_CHARS:
                    continue
                if SHELL_GLOSS.match(gloss) or INFLECTION_GLOSS.match(gloss):
                    continue
                kept += 1
                yield WiktSense(word=word, pos=pos, gloss=gloss)


def glosses_by_word(
    path: Path | None = None, *, limit_entries: int | None = None
) -> dict[tuple[str, str], list[str]]:
    """
    Definitions grouped by word and part of speech.

    Keyed by both, so a noun meaning is never paired against a verb definition
    of the same spelling.
    """
    out: dict[tuple[str, str], list[str]] = {}
    for sense in iter_senses(path, limit_entries=limit_entries):
        out.setdefault((sense.word.lower(), sense.pos), []).append(sense.gloss)
    print(f"  {len(out):,} (word, pos) entries with at least one usable gloss")
    return out
